packages/quantic/force-app/main/default/lwc/quanticDocumentSuggestion/quanticDocumentSuggestion.js

```diff
@@ -0,0 +1,181 @@
+  @track suggestions = [];
+  /** @type {boolean} */
+  loading = false;
+  /** @type {CaseAssistEngine} */
+  engine;
+  /** @type {DocumentSuggestion} */
+  documentSuggestion;
+  /** @type {Function} */
+  unsubscribeDocumentSuggestion;
+  /** @type {Array<string>} */
+  openedDocuments = [];
+
+  connectedCallback() {
+    registerComponentForInit(this, this.engineId);
+    this.template.addEventListener('rating', this.onRating);
+  }
+
+  renderedCallback() {
+    initializeWithHeadless(this, this.engineId, this.initialize);
+    this.injectIdToSlots();
+  }
+
+  /**
+   * @param {CaseAssistEngine} engine
+   */
+  initialize = (engine) => {
+    this.engine = engine;
+    this.documentSuggestion =
+      CoveoHeadlessCaseAssist.buildDocumentSuggestion(engine);
+    this.unsubscribeDocumentSuggestion = this.documentSuggestion.subscribe(() =>
+      this.updateDocumentSuggestionState()
+    );
+
+    this.actions = {
+      ...CoveoHeadlessCaseAssist.loadCaseAssistAnalyticsActions(engine),
+      ...CoveoHeadlessCaseAssist.loadDocumentSuggestionActions(engine),
+    };
+
+    if (this.fetchOnInit) {
+      engine.dispatch(this.actions.fetchDocumentSuggestions());
+    }
+  };
+
+  disconnectedCallback() {
+    this.unsubscribeDocumentSuggestion?.();
+    this.template.removeEventListener('rating', this.onRating);
+  }
+
+  updateDocumentSuggestionState() {
+    this.suggestions =
+      this.documentSuggestion.state.documents
+        .map((suggestion) => {
+          return {
+            ...suggestion,
+            raw: suggestion.fields,
+            uri: suggestion.fields.uri,
+          };
+        })
+        .slice(0, Math.max(1, this.maxDocuments)) ?? [];
+    this.openFirstDocuments();
+    this.loading = this.documentSuggestion.state.loading;
+  }
+
+  onRating = (evt) => {
+    this.engine.dispatch(
+      this.actions.logDocumentSuggestionRating(evt.detail.id, evt.detail.score)
+    );
+  };
+
+  handleSectionClick(evt) {
+    const accordion = this.template.querySelector('lightning-accordion');
+    if (
+      JSON.stringify(this.openedDocuments) !==
+      // @ts-ignore
+      JSON.stringify(accordion.activeSectionName)
+    ) {
+      if (!this.openedDocuments.includes(evt.target.name)) {
+        this.engine.dispatch(
+          this.actions.logDocumentSuggestionClick(evt.target.dataset.id)
+        );
+      }
+      // @ts-ignore
+      this.openedDocuments = accordion.activeSectionName;
+    }
+  }
+
+  openFirstDocuments() {
+    if (this.suggestions.length) {
+      this.openedDocuments = this.suggestions
+        .slice(0, Math.max(0, Number(this.numberOfAutoOpenedDocuments)))
+        .map((suggestion) => {
+          return suggestion.title;
+        });
+    }
+  }
+
+  stopPropagation(evt) {
+    evt.stopPropagation();
+  }
+
+  injectIdToSlots() {
+    const slots = this.template.querySelectorAll('slot');
+    slots.forEach((slot) => {
+      let slotContent = slot;
+      while (slotContent?.tagName === 'SLOT') {
```

Comment:
Sure! This manipulation is made to give the slots of this components access to the id of the different documents.
For example we have the slot "rating" made to show the rating of each document suggestion, it can be used this way: 
```
<c-quantic-document-suggestion>
    <c-my-custom-component-to-display-rating slot="rating"></c-my-custom-component-to-display-rating>
</c-quantic-document-suggestion>
```
The component passed as a slot here is meant to be used to display the rating for every document suggestion, the problem is that this component needs the id of the all the respective document suggestions to be able to fetch the data about the rating from somewhere. and since this component is passed as slot it has no access to this data, this data is present in quanticDocumentSuggestion component but not in the component passed as slot.

To surpass this, I'm adding this manipulation where from the quanticDocumentSuggestion I select the slots and I access the content of the slots using the method `assignedNodes` that returns a sequence of  the nodes assigned to this slot, I select the first one of this sequence which is the parent node in the slot content and I add to it a data attribute containing the data that the slot content will need to do the work correctly.
So that's what I'm doing here from the quanticDocumentSuggestion I'm adding a data attribute containing the id of the document suggestions corresponding to that slot. so if we continue with the code example I added above, the component myCustomComponentToDisplayRating will receive a data attribute containing the information needed which is the id of the document suggestions and this will allow the component to fetch the rating needed correctly.

Reply:
Thanks for the detailed explanation! My question should have been a bit more precise because what I didn't really understand was the while loop. I see your iterating until the child is no longer a slot. Do you know if it's common for this loop to run more than once? Just curious at this point.
I guess another way of doing the same could be:
`const slotContent = slot.assignedNodes.find((node) => node.tagName !== 'SLOT')`

Reply:
The loop is needed when the slots passed to a component was also a slot passed 
The loop is needed for cases like this one
We have the slot "quickview-footer" which is used to pass a custom component that will be displayed in the footer of the quickview modal, it will be used this way:
```
<c-quantic-document-suggestion>
    <c-my-custom-footer slot="quickview-footer"></c-my-custom-footer>
</c-quantic-document-suggestion>
```
and inside quanticDocumentSuggestion we have something like this:
```
// quanticDocumentSuggestion.js
<template>
...
<c-quantic-result-quickview>
    <slot name="quickview-footer" slot="footer"></slot>
</c-quantic-result-quickview>
...
</template>
```
it's a bit confusing but here myCustomFooter is the component that needs the id information in his data attribute, so here the loop is needed because from quanticResultQuickview when we try access the content of slot "footer" to add the id data attribute we find ourselves access the slot of quanticDocumentSuggestion and here we need another loop to really reach the real custom component that was sent as a slot.
In other words the loop is made for the case when the slot of a component was also a slot coming from a higher component .

Reply:
That makes sense to me. Thanks!